### DISCUSSION TITLE
Question about scaling in docker swarm/ kubernetes

### DISCUSSION THREAD

**Discussion by DanielChuDC** [fastapi-users/fastapi-users#725] [General]
Hi there, would like to ask if anyone has done scaling this library and did some stress test?

Will scaling impact or affect the performance of the backend with this library?

**Comment by frankie567** [comment]
Performance of FastAPI Users is mainly tied to the one of FastAPI and of your underlying database. We are not doing things very expensive, mostly read/writes from the database. If your database scales correctly, everything should go smoothly 😄

**Reply by DanielChuDC** [reply]
thank you @frankie567 for your reply. 

I am searching for a test with [locust](https://locust.io) with this library against SQL/NOSQL library. 

If there is no such test, probably post the result here when I do the test myself.

**Comment by junoriosity** [comment]
@frankie567 I am wondering what happens, if one scales the deployment of the API to two or more replicas of the FastAPI docker container. Would it mean a problem to the JWT token?

I ask this because I cannot find some token table in my Postgresql DB, so I assume it is somewhere stored in the container, and if there are two containers (A and B), it might be that I get my token from A (and it is stored there), but my next call gets forwarded to B, which does not know this token.

Hence, it would be good to understand where the tokens reside and whether the aforementioned concerns are realistic or not.

**Reply by frankie567** [reply]
The main benefit of JWT is that they bear all the authentication data by themselves (user id., lifetime, etc.). It's safe because they are cryptographically signed with a secret (the one you set when you instantiate the `JWTAuthentication` class). The server is then able to validate this is a legitimate token by computing the signature and checking the remaining lifetime.

Therefore, there is no need of a database to store the token and there is no problem when there are several server instances. You can read more about JWT here: https://jwt.io/introduction

The main drawback of this approach however is that there is no easy way to invalidate the JWT on the server side. With the traditional database approach, all we need to do is to remove the token from the database table. With a JWT, we have to wait it expires by itself or we need to change the server secret to invalidate every sessions. I talk about it here: https://github.com/fastapi-users/fastapi-users/discussions/786#discussioncomment-1601101

**Comment by junoriosity** [comment]
@frankie567 Many thanks for enlightening me. I am no expert on JWT. Hence, this helps me a lot.
